google/cloud/bigquery/dataset.py

```diff
@@ -145,34 +288,139 @@ class AccessEntry(object):
 
     def __init__(self, role=None, entity_type=None, entity_id=None):
         self._properties = {}
-        if entity_type in ("view", "routine", "dataset"):
-            if role is not None:
-                raise ValueError(
-                    "Role must be None for a %r. Received "
-                    "role: %r" % (entity_type, role)
-                )
-        else:
-            if role is None:
-                raise ValueError(
-                    "Role must be set for entity " "type %r" % (entity_type,)
-                )
-        self._role = role
+        if entity_type is not None:
+            self._properties[entity_type] = entity_id
+        self._role = self._properties["role"] = role
         self._entity_type = entity_type
         self._entity_id = entity_id
 
     @property
-    def role(self):
-        """str: The role of the entry."""
-        return self._role
+    def role(self) -> Optional[str]:
+        """The role of the entry."""
+        return self._properties["role"]
+
+    @role.setter
+    def role(self, value):
+        self._properties["role"] = value
+
+    @property
+    def dataset(self) -> Optional[dict]:
+        """API resource representation of a dataset reference."""
+        return self._properties["dataset"]
+
+    @dataset.setter
+    def dataset(self, value):
+        if self._role is not None:
+            raise ValueError(
+                "Role must be None for a dataset. Current " "role: %r" % (self._role)
+            )
+        if not isinstance(value, dict):
+            if isinstance(value, str):
+                value = DatasetReference.from_string(value)
+
+            if isinstance(value, (Dataset, DatasetListItem)):
+                value = value.reference
+
+            value = value.to_api_repr()
+
+        prop = {
+            "dataset": value,
+            "targetTypes": self._properties.get("targetTypes", None),
+        }
+        self._properties["dataset"] = prop
+
+    @property
+    def target_types(self) -> Optional[List[str]]:
+        """Which resources that the dataset in this entry applies to."""
+        return self._properties["dataset"]["targetTypes"]
```

Comment:
We have a helper (https://github.com/googleapis/python-bigquery/blob/8712b6f4f6c91b0f5aca3de98efe81deb2c7262f/google/cloud/bigquery/_helpers.py#L722) to make getting sub-properties safer as well.

```suggestion
        return _helpers._get_sub_prop(self._properties, ["dataset", "targetTypes"])
```